subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/SnapshotHierarchy.java

```diff
@@ -61,13 +62,9 @@
     @CheckReturnValue
     SnapshotHierarchy empty();
 
-    void visitSnapshotRoots(SnapshotVisitor snapshotVisitor);
+    Stream<FileSystemLocationSnapshot> snapshotRoots();
```

Comment:
Fully agree. I also expected the transition to be more costly, but it was really simple.

I wish we could use them in more places in `:files`, but that's still Java 6 :/. I wish we had algebraic types in Java, so we could use streams instead of `FileSystemLocationSnapshotVisitor`, too.

Anyway, one step at a time...